torchtext/csrc/gpt2_bpe_tokenizer.cpp

```diff
@@ -61,28 +63,94 @@ std::vector<std::string> gpt2_bpe_pre_tokenizer(std::string input) {
+        switch (token[pos]) {
+          case '|':
+            token.replace(pos, 1, "\\|");
+            pos += 2;
+            break;
+          case '[':
+            token.replace(pos, 1, "\\[");
+            pos += 2;
+            break;
+          case ']':
+            token.replace(pos, 1, "\\]");
+            pos += 2;
+            break;
         }
-        if (token[token.length() - 1] == ' ') { // last char is space
-          prepend_space = true;
-        } else { // push last whitespace char as a token if it is not a space
-          tokens.push_back(token.substr(token.length() - 1));
+      }
+      if (pattern.length() != 0)
+        pattern += "|";
+      pattern += token;
+    }
+
+    // break input into non-special and special parts
+    std::regex rx(pattern);
+    int64_t last_idx = 0;
+    for (auto it = std::sregex_iterator(input.begin(), input.end(), rx);
+         it != std::sregex_iterator();
+         ++it) {
+      if (it->position() > last_idx) {
+        if (isspace(input[it->position() - 1])) {
+          // lstrip
+          index_matches.push_back(
+              input.substr(last_idx, it->position() - last_idx - 1));
+        } else {
+          index_matches.push_back(
+              input.substr(last_idx, it->position() - last_idx));
         }
       }
-    } else if (prepend_space) {
-      tokens.push_back(" " + token);
-      prepend_space = false;
-    } else {
-      tokens.push_back(token);
+      index_matches.push_back(input.substr(it->position(), it->length()));
+      last_idx = it->position() + it->length() + 1;
+      if (isspace(input[last_idx])) {
+        // rstrip
+        last_idx++;
+      }
+    }
+    if (last_idx < input.length() - 1)
+      index_matches.push_back(
+          input.substr(last_idx, input.length() - last_idx));
+  } else {
+    index_matches.push_back(input);
+  }
+
+  for (std::string index_token : index_matches) {
+    bool is_never_split_token =
+        bpe_never_split_set_.find(index_token) != bpe_never_split_set_.end();
+    if (is_never_split_token) {
+      tokens.push_back(index_token);
+      continue;
+    }
+    re2::StringPiece inp(index_token);
+    while (kGPT2Regex.FindAndConsume(&inp, &token)) {
+      if (is_whitespace(token)) {
+        prepend_space = false;
+        if (inp.empty()) { // token is last token
+          tokens.push_back(token);
+        } else {
+          if (token.length() > 1) {
+            tokens.push_back(token.substr(0, token.length() - 1));
+          }
+          if (token[token.length() - 1] == ' ') { // last char is space
+            prepend_space = true;
+          } else { // push last whitespace char as a token if it is not a space
+            tokens.push_back(token.substr(token.length() - 1));
+          }
+        }
+      } else if (prepend_space) {
+        tokens.push_back(" " + token);
+        prepend_space = false;
+      } else {
+        tokens.push_back(token);
+      }
```

Comment:
I'm having a bit of a hard time following all of the additional logic to this method. Having a more detailed explanation via code comments + adding a few lines to the commented pseudocode would really help with readability for someone that is new to our codebase.

Reply:
Sorry, for the confusion. I agree that the logic is complex to understand just by looking at the code. I added some detailed explanation and example on the steps in the latest commit. Let me know if that is sufficient or any further details would help.

Reply:
I think the explanation you added really helps with the code readability, thanks!